File: apps/backend/app/schemas/transaction.py
"""
Pydantic schemas for transactions.
"""
from pydantic import BaseModel, Field, model_validator
from typing import Optional
from datetime import datetime
from app.models.transaction import TransactionState


class TransactionListingInfo(BaseModel):
    """Minimal listing info for transaction response"""
    id: int
    title: str
    category: str
    platform: str
    price_usd: int
    
    class Config:
        from_attributes = True


class TransactionCreate(BaseModel):
    """Schema for creating a transaction (initiating purchase)"""
    listing_id: int = Field(..., description="Listing ID to purchase")


class TransactionResponse(BaseModel):
    """Schema for transaction response"""
    id: int
    listing_id: int
    buyer_id: int
    seller_id: int
    amount_usd: int  # Legacy field (for backward compatibility)
    # Multi-currency support
    currency: Optional[str] = None  # Transaction currency (KSH, USD, EUR, GBP, CAD)
    amount: Optional[int] = None  # Amount in smallest unit of currency
    payment_currency: Optional[str] = None  # Payment currency (may differ from listing currency)
    payment_amount: Optional[int] = None  # Amount in payment currency
    state: TransactionState
    paystack_reference: Optional[str] = None
    paystack_authorization_code: Optional[str] = None
    paystack_customer_code: Optional[str] = None
    # Step-locked purchase flow timestamps
    purchase_initiated_at: Optional[str] = None
    payment_pending_at: Optional[str] = None
    funds_held_at: Optional[str] = None
    temporary_access_granted_at: Optional[str] = None
    verification_window_started_at: Optional[str] = None
    verification_deadline: Optional[str] = None
    account_verified_at: Optional[str] = None
    ownership_agreement_pending_at: Optional[str] = None
    ownership_agreement_signed_at: Optional[str] = None
    funds_release_pending_at: Optional[str] = None
    funds_released_at: Optional[str] = None
    completed_at: Optional[str] = None
    refunded_at: Optional[str] = None
    cancelled_at: Optional[str] = None
    # Legacy fields (for backward compatibility, mapped from new fields)
    contract_signed_at: Optional[str] = Field(default=None)  # Maps to ownership_agreement_signed_at
    credentials_released_at: Optional[str] = Field(default=None)  # Maps to temporary_access_granted_at
    access_confirmed_at: Optional[str] = Field(default=None)  # Maps to account_verified_at
    # Payout details
    payout_reference: Optional[str] = None
    commission_usd: Optional[int] = None
    payout_amount_usd: Optional[int] = None
    created_at: datetime
    # Listing relationship (optional, only included if loaded)
    listing: Optional[TransactionListingInfo] = None
    
    @model_validator(mode='before')
    @classmethod
    def set_legacy_fields(cls, data):
        """Set legacy fields from new fields before validation"""
        if isinstance(data, dict):
            # Map new fields to legacy fields if legacy fields are missing
            if 'contract_signed_at' not in data and 'ownership_agreement_signed_at' in data:
                data['contract_signed_at'] = data.get('ownership_agreement_signed_at')
            if 'credentials_released_at' not in data and 'temporary_access_granted_at' in data:
                data['credentials_released_at'] = data.get('temporary_access_granted_at')
            if 'access_confirmed_at' not in data and 'account_verified_at' in data:
                data['access_confirmed_at'] = data.get('account_verified_at')
        return data
    
    @model_validator(mode='after')
    def populate_legacy_fields(self):
        """Populate legacy fields from new fields after validation"""
        # Map new fields to legacy fields if legacy fields are None
        if not self.contract_signed_at and self.ownership_agreement_signed_at:
            object.__setattr__(self, 'contract_signed_at', self.ownership_agreement_signed_at)
        if not self.credentials_released_at and self.temporary_access_granted_at:
            object.__setattr__(self, 'credentials_released_at', self.temporary_access_granted_at)
        if not self.access_confirmed_at and self.account_verified_at:
            object.__setattr__(self, 'access_confirmed_at', self.account_verified_at)
        
        # Ensure currency fields are strings (enums get serialized)
        if hasattr(self, 'currency') and self.currency:
            if hasattr(self.currency, 'value'):
                object.__setattr__(self, 'currency', self.currency.value)
        if hasattr(self, 'payment_currency') and self.payment_currency:
            if hasattr(self.payment_currency, 'value'):
                object.__setattr__(self, 'payment_currency', self.payment_currency.value)
        
        return self
    
    class Config:
        from_attributes = True


class TransactionDetailResponse(TransactionResponse):
    """Detailed transaction response"""
    buyer_confirmed_access: bool
    notes: Optional[str]
    payment_authorization_url: Optional[str] = None  # Paystack authorization URL
    
    class Config:
        from_attributes = True

File: apps/frontend/types/transaction.ts
/**
 * Transaction Types
 * Matching backend schemas
 */

export type TransactionState =
  // STEP 1: Initiate Purchase
  | 'purchase_initiated'
  // STEP 2: Payment
  | 'payment_pending'
  | 'funds_held'
  // STEP 3: Temporary Access
  | 'temporary_access_granted'
  // STEP 4: Verification Window
  | 'verification_window'
  // STEP 5: Ownership Agreement
  | 'ownership_agreement_pending'
  | 'ownership_agreement_signed'
  // STEP 6: Final Confirmation
  | 'funds_release_pending'
  | 'funds_released'
  // STEP 7: Transaction Closed
  | 'completed'
  // Terminal states
  | 'refunded'
  | 'disputed'
  | 'cancelled'
  // Legacy states (for backward compatibility)
  | 'pending'
  | 'contract_signed'
  | 'credentials_released';

export interface Transaction {
  id: number;
  listing_id: number;
  buyer_id: number;
  seller_id: number;
  amount_usd: number; // in cents
  state: TransactionState;
  paystack_reference: string | null;
  paystack_authorization_code: string | null;
  funds_held_at: string | null;
  contract_signed_at: string | null;
  credentials_released_at: string | null;
  completed_at: string | null;
  refunded_at: string | null;
  buyer_confirmed_access: boolean;
  notes: string | null;
  created_at: string;
  updated_at: string;
  listing?: {
    id: number;
    title: string;
    category: string;
    platform: string;
    price_usd: number;
  };
  contract?: {
    id: number;
    pdf_url: string;
    signed_by_name: string | null;
    signed_at: string | null;
  };
}

export interface TransactionDetail extends Omit<Transaction, 'contract' | 'listing'> {
  listing: {
    id: number;
    title: string;
    category: string;
    platform: string;
    price_usd: number;
    description: string | null;
  };
  contract?: {
    id: number;
    pdf_url: string;
    signed_by_name: string | null;
    signed_at: string | null;
  } | null;
  payment_authorization_url?: string | null; // Paystack authorization URL
}

export interface TransactionCreate {
  listing_id: number;
}

export interface ContractSignRequest {
  signed_by_name: string;
}

export interface CredentialRevealResponse {
  username: string;
  password: string;
  recovery_email: string | null;
  two_fa_secret: string | null;
  revealed_at: string;
}

